ktest/httprecorder/request_log.go

```diff
@@ -0,0 +1,239 @@
+package httprecorder
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"regexp"
+	"sort"
+	"strings"
+	"sync"
+	"time"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/yaml"
+)
+
+type LogEntry struct {
+	Request  Request  `json:"request,omitempty"`
+	Response Response `json:"response,omitempty"`
+	Error    string   `json:"error,omitempty"`
+}
+
+type Request struct {
+	Method string      `json:"method,omitempty"`
+	URL    string      `json:"url,omitempty"`
+	Header http.Header `json:"header,omitempty"`
+	Body   string      `json:"body,omitempty"`
+}
+
+type Response struct {
+	Status     string      `json:"status,omitempty"`
+	StatusCode int         `json:"statusCode,omitempty"`
+	Header     http.Header `json:"header,omitempty"`
+	Body       string      `json:"body,omitempty"`
+}
+
+func (e *LogEntry) FormatHTTP() string {
+	var b strings.Builder
+	b.WriteString(e.Request.FormatHTTP())
+	b.WriteString(e.Response.FormatHTTP())
+	return b.String()
+}
+
+func (r *Request) FormatHTTP() string {
+	var b strings.Builder
+	b.WriteString(fmt.Sprintf("%s %s\n", r.Method, r.URL))
+	var keys []string
+	for k := range r.Header {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+	for _, k := range keys {
+		for _, v := range r.Header[k] {
+			b.WriteString(fmt.Sprintf("%s: %s\n", k, v))
+		}
+	}
+	b.WriteString("\n")
+	if r.Body != "" {
+		b.WriteString(r.Body)
+		b.WriteString("\n\n")
+	}
+	return b.String()
+}
+
+func (l *RequestLog) ReplaceTimestamp() {
+	l.mutex.Lock()
+	defer l.mutex.Unlock()
+
+	for _, entry := range l.entries {
+		entry.Request.Body = resetTimestamp(entry.Request.Body)
+		entry.Response.Body = resetTimestamp(entry.Response.Body)
+	}
+}
+
+func resetTimestamp(body string) string {
+	if body == "" {
+		return body
+	}
+	var u *unstructured.Unstructured
+	if err := yaml.Unmarshal([]byte(body), &u); err != nil {
+		return body
+	}
+
+	if u.Object["status"] == nil {
+		return body
+	}
+	status := u.Object["status"].(map[string]interface{})
+	if status["conditions"] == nil {
+		return body
+	}
+	conditions := status["conditions"].([]interface{})
+	for _, condition := range conditions {
+		cond := condition.(map[string]interface{})
+		// Use a fixed timestamp for golden tests.
+		t := time.Date(2022, time.January, 1, 0, 0, 0, 0, time.UTC)
+		cond["lastTransitionTime"] = t.Format("2006-01-02T15:04:05Z07:00")
+	}
+	b, _ := json.Marshal(u)
+	return string(b)
+}
+
+func (r *Response) FormatHTTP() string {
+	var b strings.Builder
+	b.WriteString(fmt.Sprintf("%s\n", r.Status))
+	var keys []string
+	for k := range r.Header {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+	for _, k := range keys {
+		for _, v := range r.Header[k] {
+			b.WriteString(fmt.Sprintf("%s: %s\n", k, v))
+		}
+	}
+	b.WriteString("\n")
+	if r.Body != "" {
+		b.WriteString(r.Body)
+		b.WriteString("\n")
+	}
+	return b.String()
+}
+
+type RequestLog struct {
+	// We have seen dropped logs from concurrent requests
```

Comment:
```suggestion
	// Avoid risking dropped logs from concurrent requests
```
The previous comment makes it a little ambiguous if we've seen dropped logs without this mutex (i.e. the comment explains why we have it) or if we've seen it _despite_ the mutext (implying there might be a concurrency bug somewhere that merits more investigation at some point).

This suggestion is assuming the former is the right interpretation.